From 2ba9acb2d544991bc3a8b5aa1a0efd346ddfc8ac Mon Sep 17 00:00:00 2001
From: Jungo Lin <jungo.lin@mediatek.com>
Date: Fri, 28 May 2021 11:48:51 +0800
Subject: [PATCH] [ALPS05706584] ubsan: enable AEE DB for ubsan

Trigger AEE kernel warning DB when UBSAN issue is happened.

Change-Id: I6790f211840e2f9bea2de16684aaee81b2e8733c
CR-Id: ALPS05706584
Feature: UBSAN
Signed-off-by: Jungo Lin <jungo.lin@mediatek.com>
---

diff --git a/lib/ubsan.c b/lib/ubsan.c
index f007a40..d56607c 100644
--- a/lib/ubsan.c
+++ b/lib/ubsan.c
@@ -17,6 +17,9 @@
 
 #include "ubsan.h"
 
+#include "../kernel/sched/sched.h"
+#include "../drivers/misc/mediatek/include/mt-plat/aee.h"
+
 const char *type_check_kinds[] = {
 	"load of",
 	"store to",
@@ -151,11 +154,24 @@
 
 static void ubsan_epilogue(void)
 {
+	int cpu;
+	struct rq *rq;
+
 	dump_stack();
 	pr_err("========================================"
 		"========================================\n");
 
 	current->in_ubsan--;
+	cpu = raw_smp_processor_id();
+	rq = cpu_rq(cpu);
+	if (!raw_spin_is_locked(&rq->lock)) {
+		/* AEE Kernel API Dump for UBSan */
+		aee_kernel_warning_api(__FILE__, __LINE__, DB_OPT_DEFAULT,
+			"UBSan error",
+			"[UBSan report]");
+	} else {
+		BUG();
+	}
 }
 
 static void handle_overflow(struct overflow_data *data, void *lhs,
